src/modules/csm/state.py

```diff
@@ -75,13 +94,50 @@ def status(self) -> None:
 
         logger.info(
             {
-                "msg": f"Processed {len(self.processed_epochs)} epochs",
+                "msg": f"Left {len(self.epochs_to_process)} epochs to process",
                 "assigned": assigned,
                 "included": included,
                 "avg_perf": self.avg_perf,
             }
         )
 
+    def validate(self, l_epoch: EpochNumber, r_epoch: EpochNumber) -> None:
+        if self.l_epoch > 0 and self.l_epoch != l_epoch:
+            logger.warning({"msg": f"Invalid left state border: should be {l_epoch=} but {self.l_epoch=}"})
+            raise InvalidStateLeftBorder()
+
+        if self.r_epoch > 0 and self.r_epoch != r_epoch:
+            logger.warning({"msg": f"Invalid right state border: should be {r_epoch=} but {self.r_epoch=}"})
+            raise InvalidStateRightBorder()
+
+    def validate_and_adjust(self, l_epoch: EpochNumber, r_epoch: EpochNumber):
+
+        def _discard():
+            logger.warning({"msg": "Discarding invalidated state cache"})
+            self.clear()
+            self.commit()
+
+        try:
+            self.validate(l_epoch, r_epoch)
+        except InvalidStateLeftBorder:
+            _discard()
+        except InvalidStateRightBorder:
+            if self.r_epoch < r_epoch:
+                _discard()
+            else:
+                logger.warning({"msg": "The last report was missed. Reuse the state with the new right border"})
+                # expand the state to the new right border
+                self.epochs_to_process.update(sequence(self.r_epoch - 1, r_epoch))
```

Comment:
`self.r_epoch` as well as `self.r_epoch - 1` could be already processed, so `self.r_epoch + 1`, I think.